src/ocean/net/ssl/SslClientConnection.d

```diff
@@ -0,0 +1,622 @@
+import ocean.transition;
+
+
+
+/***************************************************************************
+
+    Global context used for all SSL connections
+
+***************************************************************************/
+
+private SSL_CTX * globalSslContext;
+
+
+
+/***************************************************************************
+
+    Class representing a single SSL client connection.
+    Can only be called from inside a Task.
+
+***************************************************************************/
+
+public class SslClientConnection
+{
+    private import core.sys.posix.netdb : AF_UNSPEC, SOCK_STREAM;
+
+    static import ocean.core.ExceptionDefinitions;
+
+    private import ocean.io.select.protocol.task.TaskSelectClient;
+
+    private import ocean.sys.socket.AddrInfo;
+
+    private import ocean.sys.socket.model.ISocket;
+
+    private import ocean.sys.Epoll: epoll_event_t;
+
+    private alias epoll_event_t.Event Event;
+
+
+    /***************************************************************************
+
+        Exception class thrown on errors.
+
+    ***************************************************************************/
+
+    public static class SslException :
+        ocean.core.ExceptionDefinitions.IOException
+    {
+        import ocean.core.Exception: ReusableExceptionImplementation;
+
+        mixin ReusableExceptionImplementation!() ReusableImpl;
+
+        /*******************************************************************
+
+            Sets the exception instance.
+
+            Params:
+                file_path = path of the file
+                func_name = name of the method that failed
+                msg = message description of the error
+                file = file where exception is thrown
+                line = line where exception is thrown
+
+        *******************************************************************/
+
+        public typeof(this) set ( cstring host_path,
+                istring func_name,
+                cstring msg,
+                istring file = __FILE__, long line = __LINE__)
+        {
+            this.error_num = error_num;
+            this.func_name = func_name;
+
+            this.ReusableImpl.set(this.func_name, file, line);
+
+            this.ReusableImpl.append(": ");
+            this.ReusableImpl.append(this.func_name);
+
+            this.ReusableImpl.append(msg);
+
+            this.ReusableImpl.append(" on ").append(host_path);
+
+            return this;
+        }
+    }
+
+
+    /***************************************************************************
+
+        A minimal implementation of ISocket
+
+    ***************************************************************************/
+
+    private static class SimpleSocket : ISocket
+    {
+        public this ()
+        {
+            super (addrinfo.sizeof);
+        }
+
+        public override void formatInfo ( ref char[] buf, bool io_error )
```

Comment:
Needs a comment why having empty impl is OK here

Reply:
I can't answer that. I don't know why that function is abstract in the first place. I shouldn't have to declare this class at all.

Reply:
@gavin-norman-sociomantic @david-eckardt-sociomantic ?

Reply:
So you should provide simple human readable string here for logging purposes.

Reply:
Can't there be a default implementation? It's not of any interest for this module. It seems crazy to have to provide a derived class, just for debugging!

Reply:
Yeah, the default implementation could be:

```
        sformat(buf, "fd={}, ioerr={}",
            this.fileHandle, io_error);
```